---
title: 16. References
last_updated: Thu Nov 21 16:49:12 2019
sidebar: mydoc_sidebar
permalink: mydoc_systemPipeRIBOseq_16.html
---
 
1. Aspden, Julie L, Eyre-Walker, Ying Chen, Phillips, Rose J, Amin, Unum, Mumtaz, Muhammad Ali S, Brocard, Michele, Couso, Juan-Pablo (2014) Extensive translation of small Open Reading Frames revealed by Poly-Ribo-Seq. Elife, 3: e03528; [URL](http://dx.doi.org/10.7554/eLife.03528)
2. Girke, Thomas (2014) systemPipeR: NGS workflow and report generation environment. NA, NA: NA; [URL](https://github.com/tgirke/systemPipeR)
3. Ingolia, N T, Ghaemmaghami, S, Newman, J R, Weissman, J S (2009) Genome-wide analysis in vivo of translation with nucleotide resolution using ribosome profiling. Science, 324: 218--223; [URL](http://dx.doi.org/10.1016/j.ymeth.2009.03.016)
4. Ingolia, N T, Lareau, L F, Weissman, J S (2011) Ribosome profiling of mouse embryonic stem cells reveals the complexity and dynamics of mammalian proteomes. Cell, 147: 789--802; [URL](http://www.ncbi.nlm.nih.gov/pubmed/22056041)
5. Juntawong, Piyada, Girke, Thomas, Bazin, J\'{e}r\'{e}mie, Bailey-Serres, Julia (2014) Translational dynamics revealed by genome-wide profiling of ribosome footprints in Arabidopsis. Proc. Natl. Acad. Sci. U. S. A., 111: E203--12; [URL](http://dx.doi.org/10.1073/pnas.1317811111)
6. Juntawong, Piyada, Hummel, Maureen, Bazin, Jeremie, Bailey-Serres, Julia (2015) Ribosome profiling: a tool for quantitative evaluation of dynamics in mRNA translation. Methods Mol. Biol., 1284: 139--173; [URL](http://dx.doi.org/10.1007/978-1-4939-2444-8_7)
7. Kim, Daehwan, Langmead, Ben, Salzberg, Steven L (2015) HISAT: a fast spliced aligner with low memory requirements. Nat. Methods, 12: 357--360; [URL](NA)
8. Kim, Daehwan, Pertea, Geo, Trapnell, Cole, Pimentel, Harold, Kelley, Ryan, Salzberg, Steven L (2013) TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions. Genome Biol., 14: R36; [URL](http://dx.doi.org/10.1186/gb-2013-14-4-r36)
9. Langmead, Ben, Salzberg, Steven L (2012) Fast gapped-read alignment with Bowtie 2. Nat. Methods, 9: 357--359; [URL](http://dx.doi.org/10.1038/nmeth.1923)
10. Lawrence, Michael, Huber, Wolfgang, Pag\`{e}s, Herv\'{e}, Aboyoun, Patrick, Carlson, Marc, Gentleman, Robert, Morgan, Martin T, Carey, Vincent J (2013) Software for computing and annotating genomic ranges. PLoS Comput. Biol., 9: e1003118; [URL](http://dx.doi.org/10.1371/journal.pcbi.1003118)
11. Robinson, M D, McCarthy, D J, Smyth, G K (2010) edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26: 139--140; [URL](http://dx.doi.org/10.1093/bioinformatics/btp616)
<br><br><center><a href="mydoc_systemPipeRIBOseq_15.html"><img src="images/left_arrow.png" alt="Previous page."></a>Previous Page &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Next Page
<a href="mydoc_systemPipeRIBOseq_01.html"><img src="images/right_arrow.png" alt="Next page."></a></center>
